tests/FSharp.Compiler.ComponentTests/Conformance/Types/UnionTypes/UnionStructTypes.fs

```diff
@@ -4,6 +4,76 @@ open Xunit
 open FSharp.Test.Compiler
 
 module UnionStructTypes =
+    
+    [<Fact>]
+    let ``Union struct can share fields with equal name and type`` () =
+        Fsx """
+module FieldSharingTest
+open Microsoft.FSharp.Data.UnitSystems.SI.UnitNames
+[<Struct>]
+type StructUnion =
+    | A of int64*int64*int64
+    | B of int64*int64*int64
+    | C of int64*int64*int64
+    | D of int64*int64*int64
+    | E of int64*int64*int64
+    | F of int64*int64*int64
+
+type MyLittleAbbreviation = int64
+type DoubleAbbrev = MyLittleAbbreviation
+[<Measure>]
+type MyUnit
+
+[<Struct>]
+type SharingAfterErasure =
+    | Length of int64<meter>
+    | Time of int64<second>
+    | Temperature of int64<kelvin>
+    | Pressure of int64<pascal>
+    | MyLittle of MyLittleAbbreviation
+    | DoublyAbbreviated of DoubleAbbrev
+    | JustPlain of int64
+    | MyUnit of int64<MyUnit>
+
+[<Struct>]
+type GenericStruct<'T> = 
+    | HasOne of field1:'T
+    | HasTwo of field1:'T * field2:'T
+    | HasThree of field1:'T * field2:'T * field3:'T
+
+[<Struct>]
+type MixingFields<'T> = 
+    | JustInt of sField:string
+    | JustString of intField:int
+    | IntAndString of sField:string * intField:int
+    | StringAndT of sField:string * tField:'T
+    | Nothing
+    | AllThree of sField:string * intField:int * tField:'T
+
+let structUnionSize = sizeof<StructUnion>
+let genericSizeForInt = sizeof<GenericStruct<int>>
+let genericSizeForString = sizeof<GenericStruct<string>>
+let sizeForMixingWithBool = sizeof<MixingFields<bool>>
+let sizeForMixingWithString = sizeof<MixingFields<string>>
+let sizeForSharingAfterErasure = sizeof<SharingAfterErasure>
+
+let equalityMustKeepWorking() =
+    if SharingAfterErasure.Length(15L<meter>) = SharingAfterErasure.JustPlain(15L) then
+        failwith "Erasure must not kill equality for UoMs!"
+
+do equalityMustKeepWorking()
+
+[<EntryPoint>]
+let main _args = 
+    printf "BasicThreeLongs=%i;GenericOfInt=%i;GenericOfString=%i;MixWithBool=%i;MixWithString=%i;Erasure=%i" structUnionSize genericSizeForInt genericSizeForString sizeForMixingWithBool sizeForMixingWithString sizeForSharingAfterErasure
+    0
+        """   
+        |> withLangVersionPreview
+        |> asExe
+        |> compile
+        |> shouldSucceed
+        |> run        
+        |> verifyOutput "BasicThreeLongs=32;GenericOfInt=16;GenericOfString=32;MixWithBool=24;MixWithString=24;Erasure=16"
```

Comment:
It's all related to struct size, that is why it is together.
When a things goes wrong in the IL, the number of tests failing because of the size should be minimal.

Also, tests which need to `run` should be kept to a small number, since they have a massive overhead.